### ISSUE TITLE
Make benchmarks more visible

### ISSUE THREAD

**Issue by dyedgreen** [dyedgreen/deno-sqlite#14]
It would be nice to record bencharks on commit to master to a csv file and render them for display in the READMEs. (compare https://deno.land/benchmarks)

**Comment by andykais** [comment]
very curious how this library's performance compares to the infamous [better-sqlite3](https://github.com/JoshuaWise/better-sqlite3) library

**Comment by dyedgreen** [comment]
Yeah, that might be an interesting comparison. I'll have a look at seeing how they compare (although it should probably be mentioned, that this library stores the files in-memory only, with explicit save, due to limitations in WASM; resulting in an API that has to specifically save changes, similar to Python)

**Comment by andykais** [comment]
> this library stores the files in-memory only

Is this still true after https://github.com/dyedgreen/deno-sqlite/issues/49 was closed?

**Comment by dyedgreen** [comment]
No, the library now provides a custom VFS for Sqlite, which means that files are read and written by Sqlite directly. How much of a file is stored in memory is entirely up to Sqlite, just like you would get when using Sqlite e.g. as a C library 😅

**Comment by andykais** [comment]
Sweet! Fwiw, this is a benchmark comparing inserting 1,000,000 rows inside a transaction with your library vs better-sqlite3. The difference is pretty much negligible

#### deno sqlite:
```ts
import { DB } from "https://deno.land/x/sqlite/mod.ts";


const db = new DB("test.db");

db.query('DROP TABLE IF EXISTS people')
db.query(`CREATE TABLE IF NOT EXISTS people (
  name TEXT NOT NULL
)`)

const rows = Array(1000000).fill(null).map((_,i) => i)
const stmt = db.prepareQuery("INSERT INTO people (name) VALUES (?)")
console.time('insert 1,000,000 rows in a transaction')
db.query('begin')
for (const row of rows) {
  stmt([row])
}
console.timeEnd('insert 1,000,000 rows in a transaction')
console.time('select from 1,000,000 rows')
db.query('SELECT COUNT(*) FROM people').next()
db.query('SELECT * FROM people LIMIT 1').next()
console.timeEnd('select from 1,000,000 rows')
db.query('commit')
``` 
insert 1,000,000 rows in a transaction: 1435ms
select from 1,000,000 rows: 17ms



#### better-sqlite:
```js
const Sqlite = require('better-sqlite3')

const db = new Sqlite('test.db')

db.exec(`DROP TABLE IF EXISTS people;
CREATE TABLE people (
  name TEXT NOT NULL
);
`)

const rows = Array(1000000).fill(null).map((_,i) => i)
const stmt = db.prepare("INSERT INTO people (name) VALUES (?)")
console.time('insert 1,000,000 rows in a transaction')
db.transaction(() => {
for (const row of rows) {
  stmt.run(row)
}
})()
console.timeEnd('insert 1,000,000 rows in a transaction')
console.time('select from 1,000,000 rows')
db.prepare('SELECT COUNT(*) FROM people').get()
db.prepare('SELECT * FROM people LIMIT 1').get()
console.timeEnd('select from 1,000,000 rows')
```
insert 1,000,000 rows in a transaction: 1.410s
select from 1,000,000 rows: 0.786ms

**Comment by andykais** [comment]
if you have any ideas on benchmarks that would be more telling, I would be interested to hear them

**Comment by dyedgreen** [comment]
I think in your benchmark with Deno SQLite, you only close the transaction after the read (I was surprised the read was so much slower in comparison).

I got the following (although the performance is not comparable, since I used a different computer form your better sqlite3 benchmark). The read still seems to be slower, but I'm not exactly sure why; the cost probably comes from the JS based file-system interface, or from passing the string through the WASM <-> js boundary.

Very happy to see that the performance for the bulk insert is comparable though!

```typescript
import { DB } from "https://deno.land/x/sqlite/mod.ts";

const db = new DB("test.db");

db.query("DROP TABLE IF EXISTS people");
db.query(`CREATE TABLE IF NOT EXISTS people (
  name TEXT NOT NULL
)`);

const rows = Array(1000000).fill(null).map((_, i) => i);
const stmt = db.prepareQuery("INSERT INTO people (name) VALUES (?)");

/* INSERT */
console.time("insert 1,000,000 rows in a transaction");

db.query("begin");
for (const row of rows) {
  stmt([row]);
}
db.query("commit");

console.timeEnd("insert 1,000,000 rows in a transaction");

/* SELECT */
console.time("select from 1,000,000 rows");

db.query("SELECT COUNT(*) FROM people").next();
db.query("SELECT * FROM people LIMIT 1").next();

console.timeEnd("select from 1,000,000 rows");
```

```
insert 1,000,000 rows in a transaction: 514ms
select from 1,000,000 rows: 8ms
```

**Comment by andykais** [comment]
> the cost probably comes from the JS based file-system interface, or from passing the string through the WASM <-> js boundary

Do you have an idea of a benchmark that would stress this part of the library? Maybe 1,000,000 selects? I am essentially asking you to highlight the weaknesses of your library 😅 so sorry for that

**Comment by andykais** [comment]
ok wow. That did it. `better-sqlite3` was able to perform 13,000 selects in the amount of time it took this library to perform 100 selects.

#### deno sqlite:
```ts
const select_stmt = db.prepareQuery('SELECT * FROM people WHERE name = ?')
const select_count = 100
console.time(`select ${select_count} times`)
for (let i = 0; i < select_count; i++) {
  const result = select_stmt([i])
  const val = result.next()
  result.return()
}
console.timeEnd(`select ${select_count} times`)
```
select 100 times: 19253ms


#### better-sqlite3:
```js
const select_stmt = db.prepare('SELECT * FROM people WHERE name = ?')
const select_count = 13000
console.time(`select ${select_count} times`)
for (let i = 0; i < select_count; i++) {
  select_stmt.get(i)
}
console.timeEnd(`select ${select_count} times`)
```
select 13000 times: 18.610s



I'm sure the translation layer has something to do with this, but I cant help but think the biggest cost here is setting up an iterator on each select and needing to close each one...Especially for select statements that return single rows. Why use this iterator approach over simple data return? Is this supposed to be used for pagination?

**Comment by dyedgreen** [comment]
Hm yeah. That might be the problem actually. The Rows iterator is used to provide a nice interface for retrieving rows, and cleaning up the connections correctly automatically (even if there is eg an exception in a fir loop).

But as you point out, it might save a lot of performance if there was a `queryOne` method that skips the Rows iterator, and directly returns the results.

**Comment by andykais** [comment]
> The Rows iterator is used to provide a nice interface for retrieving rows, and cleaning up the connections correctly automatically

I guess I am confused why you would ever want an open connection at all? Why not just fire and forget? Does the iterator keep the un-iterated rows from being read into memory? If the use case is pagination, then in reality most applications want to read a chunk of rows into memory, then wait for the user ask to read a second chunk into memory. In those cases storing a cursor is much more reliable than storing an in memory pointer (the iterator).

Just my 2 cents, but since this library isn't an trying to be an ORM (which is a good thing), you can just let library users build their own access abstractions over simple reliable primitives (get all, get one, insert, etc)

**Comment by dyedgreen** [comment]
Currently, the Rows is basically just an SQLite statement pointer and a flag to indicate if all rows were returned. It implements the iterator protocol, where each call to next() simply retrieves the next row from the sqlite statement.

There is also the prepareQuery api, which keeps the statement for reuse (if you get the Rows from query, it will clean up the statement).

Basically, something really nice that’s enabled by this design is the following:
```javascript
for (const [name, age] of db.query(…)) {
  if (age > 10) {
    throw Error(“Bad age.”);
  }
  // … more steps …
}
```
Now when this throws, the statement will be cleaned up automatically, because JavaScript will call return on the iterator.

Basically, having it this way means the library can handle most of the memory management automatically. (The exception being prepared queries.)

**Comment by andykais** [comment]
I guess I fail to see why you would ever want the connection open for a period of time? Why is that better than:
```ts
const rows = db.query('SELECT name, age FROM people') // close immediately
for (const [name, age] of rows) {

}
```

**Comment by dyedgreen** [comment]
If I remember correctly, the idea was originally, that you could process queries of any size, without having to load everything into memory at any given time, by having queries return iterators over results.

This is also closer to how the underlying C-API works (where you can iterator over rows until you're done). But the main idea was that this approach will always 'just work', regardless of your workload.

If the construction of the iterator turns out to be the bottleneck on single row selects though, I'm very interested in providing an additional method for returning exactly one value (e.g. a `queryOne` on `DB` and a similar method on `PreparedQuery`), which would skip constructing a `Rows` class all together.

**Comment by andykais** [comment]
> If I remember correctly, the idea was originally, that you could process queries of any size, without having to load everything into memory at any given time, by having queries return iterators over results.



I feel like what you are describing is a database cursor. (e.g. [this interface here](https://github.com/JoshuaWise/better-sqlite3/blob/master/docs/api.md#iteratebindparameters---iterator)). There are definitely use cases for it, but I dont know if it should be the default retrieval method. I think a better design would be:
```ts
type Row = {[column: string]: any}
type Stats = { changes: number, last_insert_rowid: number }
interface Statement {
  run(params: any[]): Stats
  get_one(params: any[]): Row
  get_many(params: any[]): Row[]
  // or "cursor" or whatever you like
  iterate(params: any[]): Iterator
}
```
I am not being very creative here, instead stealing a lot of the interface from better-sqlite3, this may not be the best design for what you want to build. This big thing here though is that I think different db access patterns warrant different methods, for type signatures and for your connection handling logic.


I think we saw that hoisting this interface to JS code for every read operation has a significant performance impact. This extends past single row retreival. What if I want to access exactly 10 rows, or exactly 100 rows? Theres still an open connection to handle (implicitly or otherwise)

**Comment by dyedgreen** [comment]
Yes, from what I can tell the most common patterns are:
- execute (query none)
- query one
- query all
- query cursor / iterator

The current API actually supports both the execute and the query a cursor usecases (since it will not construct an iterator if a query returns no rows, and instead returns the global `Empty`.

The other two use cases are actually also covered by the current API (that’s one of the reasons the current approach is so elegant):

```typescript
// query all
const rows = […db.query("SELECT * FROM users")];
// query one/ n
const [first] = db.query("SELECT * FROM users LIMIT 1");
```

But you’re right that the construction of the iterator object might be the performance bottleneck. I think for most use-cases the current API should be fast enough and it’s very nice/ easy to use.

But for higher performance usecases, I think we should provide a faster API. I’m not 100% sure if we should make specialized functions for common use cases (i.e. queryOne and queryAll). Or if we should expose the underlying SQLite interface for users with very special requirements. 🤔

**Comment by andykais** [comment]
It is ultimately up to you. My 2 cents are that simple functions should be the first and foremost interface. They have very low performance impact, and no impact on the database. I'm not sure all the impacts of a changeset iterator (e.g. does it need to store a lot of extra info in the database if the rows being iterated over are changed in the database?). There probably aren't a lot of implications outside of performance, but there are some questions there.

Your example also has to iterate the whole result just to close the iterator, that feels like wasted work to me. What if I want to just know the number of rows returned? What if I want to walk backwards through the rows? An iterator doesnt know its own length, and its a singly linked list. I imagine most users are just going to use the iterator to populate an array before doing anything interesting with the result.

**Comment by dyedgreen** [comment]
Closing this for now, since #138 has massively improved performance.
